api/routers/auth.js

```diff
@@ -13,18 +13,38 @@ function redirectIfAuthenticated(req, res, next) {
   req.session.authenticated ? res.redirect('/') : next();
 }
 
-function onSuccess(req, res) {
-  req.session.authenticated = true;
-  req.session.authenticatedAt = new Date();
-  req.session.save(() => {
-    res.redirect(req.session.authRedirectPath || '/');
-  });
+// function onSuccess(req, res, next) {
+//   console.log('auth callback', third)
+//   req.session.authenticated = true;
+//   req.session.authenticatedAt = new Date();
+//   req.session.save(() => {
+//     res.redirect(req.session.authRedirectPath || '/');
+//   });
+// }
+
+function uaaCallback(req, res, next) {
+  // eslint-disable-next-line consistent-return
+  passport.authenticate('uaa', (err, user, info) => {
+    if (err) {
+      return res.forbidden();
+    }
+    if (user) {
+      req.session.passport = { user: user.id };
+      req.session.authenticated = true;
+      req.session.authenticatedAt = new Date();
+    } else {
+      req.session.flash = { error: [info.message] };
```

Comment:
Was passport-oauth2's built-in callback doing anything to explicitly clear a session on an undefined `user`?

Reply:
Short answer: no; also there shouldn't be any relevant session information if the user comes back undefined. Longer answer:

The default behavior of `passport.authenticate` is detailed [here](https://github.com/jaredhanson/passport/blob/master/lib/middleware/authenticate.js). Providing a callback will cause an early return for three strategy methods: `allFailed` (all provided strategies failed to authenticate), `success` and `error`.
- Default error behavior is `next(error)` , we override with `res.forbidden` (though we should probably change to 401)
- Default failure behavior is to (based on options) set flash messages, session failure messages, and redirect. We show the flash messages and redirect.
- Default success behavior is mostly based on options and then calls [`req.logIn`](https://github.com/jaredhanson/passport/blob/0575de90dc0e76c1b8ca9cc676af89bd301aec60/lib/http/request.js#L24-L46) with the user. I'm going to rewrite our callback to call this rather than interacting with the session directly

Reply:
Thanks for the detail with the deep dive.